appsec/src/helper/sampler.hpp

```diff
@@ -6,42 +6,285 @@
+
+            // should not happen... but if it does, return a fake entry
+            return {entries[Capacity], true};
+        }
+    };
+
+    bool hit(table &table, std::uint64_t number) noexcept
+    {
+        const std::uint32_t now = time_provider.now() - time_bias;
+        const std::uint32_t report_threshold = now - Threshold;
+
+    another_slot:
+        auto [entry, exists] = table.find_slot(number);
+        if (!exists) {
+            auto old_size = table.size.fetch_add(1);
+
+            if (old_size >= MaxItems) {
+                bool expected = false;
+                if (rebuild_in_progress.compare_exchange_strong(
+                        expected, true, std::memory_order_relaxed)) {
+                    std::thread([this, &table, report_threshold]() {
+                        rebuild_table(table, report_threshold);
+                    }).detach();
+                }
+            }
+            if (old_size >= Capacity) {
+                // no space to add anything
+                table.size.fetch_add(-1);
+                return false;
+            }
+
+            std::uint64_t exp_number{};
+            if (!entry.key.compare_exchange_strong(
+                    exp_number, number, std::memory_order_relaxed)) {
+                if (exp_number == number) {
+                    // another thread inserted the same number
+                    // presumably between find_slot() and the CAS only a very
+                    // small amount of time passed
+                    return false;
+                }
+                table.size.fetch_add(-1);
+                goto another_slot; // try and find another slot
+            }
+
+            typename entry::edata exp_data{};
+            typename entry::edata desired_data{now, now};
+            if (!entry.data.compare_exchange_strong(
+                    exp_data, desired_data, std::memory_order_relaxed)) {
+                // though we created the entry, another thread updated the data
+                return false;
+            }
+
+            return true;
+        }
+
+        // else there is already something
+        auto cur_data = entry.data.load(std::memory_order_relaxed);
+        if (cur_data.last_reported <= report_threshold) {
+            // potentially a hit
+            typename entry::edata desired_data{now, now};
+
+        retry:
+            if (!entry.data.compare_exchange_strong(
+                    cur_data, desired_data, std::memory_order_relaxed)) {
+                // another thread just updated it
+                // was it a hit?
+                if (cur_data.last_accessed == cur_data.last_reported) {
+                    // then this one should not be a hit
+                    return false;
+                }
+
+                // the other thread did not register a hit
+                // we retry if our idea of time is ahead
+                if (cur_data.last_accessed < now) {
+                    goto retry;
+                }
```

Comment:
I'm not so sure about this one, if you don't retry, you still return a hit (true), but you don't update...

Reply:
you're right, this is missing a return false